Sources/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -0,0 +1,456 @@
+      guard !src.starts(with: "[[:") else { return nil }
+
+      if src.tryEat("[") {
+        return src.tryEat("^") ? .inverted : .normal
+      }
+      return nil
+    }
+  }
+
+  /// Try to consume a binary operator from within a custom character class
+  ///
+  ///     CustomCCBinOp -> '--' | '~~' | '&&'
+  ///
+  mutating func lexCustomCCBinOp() throws -> Value<CustomCharacterClass.SetOp>? {
+    try recordLoc { src in
+      // TODO: Perhaps a syntax options check (!PCRE)
+      // TODO: Better AST types here
+      if src.tryEat(sequence: "--") { return .subtraction }
+      if src.tryEat(sequence: "~~") { return .intersection }
+      if src.tryEat(sequence: "&&") { return .symmetricDifference }
+      return nil
+    }
+  }
+
+  /// Consume an escaped atom, starting from after the backslash
+  ///
+  ///     Escaped          -> KeyboardModified | Builtin
+  ///                       | UniScalar | Property
+  ///
+  /// TODO: references
+  mutating func expectEscaped(
+    isInCustomCharacterClass ccc: Bool
+  ) throws -> Value<Atom> {
+    try recordLoc { src in
+      // Keyboard control/meta
+      if src.tryEat("c") || src.tryEat(sequence: "C-") {
+        return .keyboardControl(try src.expectASCII().value)
+      }
+      if src.tryEat(sequence: "M-\\C-") {
+        return .keyboardMetaControl(try src.expectASCII().value)
+      }
+      if src.tryEat(sequence: "M-") {
+        return .keyboardMeta(try src.expectASCII().value)
+      }
+
+      let char = src.eat()
+
+      // Single-character builtins
+      if let builtin = Atom.EscapedBuiltin(
+        char, inCustomCharacterClass: ccc
+      ) {
+        return .escaped(builtin)
+      }
+
+      switch char {
+      // Scalars
+      case "u", "x", "U":
+        return try .scalar(
+          src.expectUnicodeScalar(escapedCharacter: char).value)
+
+      // Unicode property checks
+      case "p", "P":
+        fatalError("TODO: properties")
+
+      case "1"..."9", "g", "k":
+        fatalError("TODO: References")
+
+      default: return .char(char)
+      }
+    }
+  }
+
+
+  /// Try to consume an Atom. 
+  ///
+  ///     Atom             -> SpecialCharacter | POSIXSet
+  ///                       | '\' Escaped | [^')' '|']
+  ///     SpecialCharacter -> '.' | '^' | '$'
+  ///     POSIXSet         -> '[[:' name ':]]'
+  ///
+  /// If `SyntaxOptions.nonSemanticWhitespace` is enabled, also accepts:
+  ///
+  ///     ModernGroupStart -> '(_:'
+  ///
+  mutating func lexAtom(
+    isInCustomCharacterClass customCC: Bool
+  ) throws -> Value<Atom>? {
+    try recordLoc { src in
+      // Check for not-an-atom, e.g. parser recursion termination
+      if src.isEmpty || src.peek() == ")" || src.peek() == "|" {
+        return nil
+      }
+      // TODO: Store customCC in the atom, if that's useful
+
+      // POSIX named set
+      if src.tryEat(sequence: "[[:") {
+        // ...
+        try src.expect(sequence: ":]]")
+        fatalError("TODO: posix named sets")
+      }
```

Comment:
POSIX char class atoms are more like `[:xxx:]`, it's just that they're only allowed in custom character classes, but you can have e.g `[ab[:space:]\d[:upper:]cd]`

Reply:
🤯